services/datamanager/datasync/upload_data_capture_file.go

```diff
@@ -26,19 +28,72 @@ func uploadDataCaptureFile(ctx context.Context, client v1.DataSyncServiceClient,
 		return nil
 	}
 
-	uploadMD := &v1.UploadMetadata{
-		PartId:           partID,
-		ComponentType:    md.GetComponentType(),
-		ComponentName:    md.GetComponentName(),
-		MethodName:       md.GetMethodName(),
-		Type:             md.GetType(),
-		MethodParameters: md.GetMethodParameters(),
-		FileExtension:    md.GetFileExtension(),
-		Tags:             md.GetTags(),
+	if md.GetType() == v1.DataType_DATA_TYPE_BINARY_SENSOR && md.GetMethodName() == "GetImages" {
+
+		if len(sensorData) > 1 {
+			return errors.New("binary sensor data file with more than one sensor reading is not supported")
+		}
+
+		// Pull timestamps out of metadata
+		treq := sensorData[0].GetMetadata().GetTimeRequested()
+		trec := sensorData[0].GetMetadata().GetTimeReceived()
+
+		// Convert SensorData into a GetImagesResponse
+		var getimgsres pb.GetImagesResponse
+		spbstruct := sensorData[0].GetStruct() // This is a GetImagesResponse, need to cast structpb to correct type
+		mp := spbstruct.AsMap()
+		if err := mapstructure.Decode(mp, &getimgsres); err != nil {
+			return nil
+		}
+
+		for _, img := range getimgsres.Images {
+			newSensorData := []*v1.SensorData{
+				{
+					Metadata: &v1.SensorMetadata{
+						TimeRequested: treq,
+						TimeReceived:  trec,
+					},
+					Data: &v1.SensorData_Binary{
+						Binary: img.GetImage(),
+					},
+				},
+			}
+			newUploadMD := &v1.UploadMetadata{
+				PartId:           partID,
+				ComponentType:    md.GetComponentType(),
+				ComponentName:    md.GetComponentName(),
+				MethodName:       md.GetMethodName(),
+				Type:             md.GetType(),
+				MethodParameters: md.GetMethodParameters(),
+				FileExtension:    getFileExtFromImageFormat(img.GetFormat()),
+				Tags:             md.GetTags(),
+			}
+			if err := helperUploadDataCaptureFile(ctx, client, newUploadMD, newSensorData, partID, f.Size()); err != nil {
+				return err
+			}
+		}
+	} else {
+		// Build UploadMetadata
+		uploadMD := &v1.UploadMetadata{
+			PartId:           partID,
+			ComponentType:    md.GetComponentType(),
+			ComponentName:    md.GetComponentName(),
+			MethodName:       md.GetMethodName(),
+			Type:             md.GetType(),
+			MethodParameters: md.GetMethodParameters(),
+			FileExtension:    md.GetFileExtension(),
+			Tags:             md.GetTags(),
+		}
+		helperUploadDataCaptureFile(ctx, client, uploadMD, sensorData, partID, f.Size())
 	}
+	return nil
+}
+
+func helperUploadDataCaptureFile(ctx context.Context, client v1.DataSyncServiceClient, uploadMD *v1.UploadMetadata, sensorData []*v1.SensorData, partID string, fileSize int64) error {
 
 	// If it's a large binary file, we need to upload it in chunks.
-	if md.GetType() == v1.DataType_DATA_TYPE_BINARY_SENSOR && f.Size() > MaxUnaryFileSize {
+	if uploadMD.GetType() == v1.DataType_DATA_TYPE_BINARY_SENSOR && fileSize > MaxUnaryFileSize {
+
 		if len(sensorData) > 1 {
```

Comment:
In order to only check this once, in `uploadDataCaptureFile` after the `len(sensorData) == 0` check, can return this error if `uploadMD.GetType() == v1.DataType_DATA_TYPE_BINARY_SENSOR && len(sensorData) > 1`

Reply:
Yep.